### ISSUE TITLE
require-atomic-updates bug

### ISSUE THREAD

**Issue by getify** [eslint/eslint#11842]
**Tell us about your environment**

* **ESLint Version:** 5.15.0
* **Node Version:** 12.2.0
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

```js
"require-atomic-updates": "error"
```

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
var fn = /* .. */;
var pr = /* .. */;
(async function IIFE(){
	fn = fn.call(await pr);
        // or: fn = fn(await pr);
})();
```

**What did you expect to happen?**

I expected no error for this rule.

**What actually happened? Please include the actual, raw output from ESLint.**

"Possible race condition: `fn` might be reassigned based on an outdated value of `fn`."

Here, the `fn` identifier appears on both sides of the `=`, and there is an `await`, but `fn`'s value on the right hand-side is not read until after the `await` has finished, therefore there is no danger of a race condition.

**Are you willing to submit a pull request to fix this bug?**

Perhaps.

**Comment by mysticatea** [comment]
Thank you for your report.

However, this is the correct behavior; the `fn` causes race condition in technically.

1. evaluate the callee and store the result. In this step, the global variable `fn`'s value was stored in memory.
2. evaluate the arguments and store the result. In this step, other processes can interrupt as `await pr` contains `await` expression.
3. evaluate the call expression with the stored results of step 1 and 2, then store the result.
4. assign the result of step 3 to the global variable `fn`.

Therefore, if the global variable `fn` was rewritten in step 2, it causes unexpected results. It's what the rule should report.

**Comment by getify** [comment]
I mis-stated in my OP. I got confused. I wasn't expecting `fn` on the right-hand side to be evaluated after the `await`. I was actually expecting the opposite, which is why this rule surprised me. Sorry about that.

Actually, this "problem" has nothing to do with `await`. This code could have the same issue:

```js
function foo() {
   fn = function(){ console.log("nope"); };
   return 42;
}

function bar(x) {
   console.log(x);
}

var fn = bar;

fn( foo() );  // 42  -- not "nope"
```

I don't think this rule would catch that situation would it?

I would find it strange if, during the course of making a function call (resolving its arguments, with or without `await`), the function-to-be-called could be swapped out. I am glad JS doesn't support that, and I am not sure why anyone would expect that to be true, because I don't think it's possible in any sense or situation.

TBH I don't know why this rule needs to warn on the function-call-being-reassigned case, since it's not  possible and thus could never have created a bug or gone against someone's expectation.

Is there any scenario, `await` or not, where someone could reasonably have expected a call-argument expression to have swapped out which function was being called? I can't think of any.

**Comment by mysticatea** [comment]
Previous discussion is here: https://github.com/eslint/eslint/issues/10405

What the rule does is very simple.

- For each variable which came from upper scopes, if it was read before `await`/`yield` and it was written after `await`/`yield`, the rule warns the variable.

In other words, that disallows read-update-write in the place that can be run in parallel.

This is a well-known problem for us in some contexts.

- Database: it's what the transaction system resolves.
- Multi-threading: it's what mutex-lock resolves.

In general, we don't address function calls as parallel processing.

I'd like to recommend to prefer parameters and return values rather than shared mutable variables in async/generator functions.

**Comment by getify** [comment]
Would you consider adding a configuration for this rule which would not have it report on cases where the variable in question is a function being called, since (as I explained above) that could never have been a bug in the first place, so this rule reporting on it is a bit of extraneous info?

I like the other more obvious benefits of this rule, but this particular scenario has caused me to turn off the rule. I'd rather have the rule turned on with this particular case configured off.

**Comment by mysticatea** [comment]
I don't think such an option is reasonable. There are no differences between the callee of function calls and other references.
